quickwit-janitor/src/actors/retention_policy_evaluator.rs

```diff
@@ -0,0 +1,460 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use itertools::Itertools;
+use quickwit_actors::{Actor, ActorContext, Handler};
+use quickwit_metastore::{IndexMetadata, Metastore};
+use tracing::{error, info};
+
+use crate::retention_policy_evaluation::run_evaluate_retention_policy;
+
+const RUN_INTERVAL: Duration = Duration::from_secs(60 * 60); // 1 hours
+
+#[derive(Clone, Debug, Default)]
+pub struct RetentionPolicyEvaluatorCounters {
+    /// The number of refresh the config passes.
+    pub num_refresh_passes: usize,
+
+    /// The number of evaluation passes.
+    pub num_evaluation_passes: usize,
+
+    /// The number of deleted splits.
+    pub num_discarded_splits: usize,
+}
+
+#[derive(Debug)]
+struct Loop;
+
+#[derive(Debug)]
+struct Evaluate {
+    index_id: String,
+}
+
+/// An actor for collecting garbage periodically from an index.
+pub struct RetentionPolicyEvaluator {
+    metastore: Arc<dyn Metastore>,
+    index_metadatas: HashMap<String, IndexMetadata>,
+    counters: RetentionPolicyEvaluatorCounters,
+}
+
+impl RetentionPolicyEvaluator {
+    pub fn new(metastore: Arc<dyn Metastore>) -> Self {
+        Self {
+            metastore,
+            index_metadatas: HashMap::new(),
+            counters: RetentionPolicyEvaluatorCounters::default(),
+        }
+    }
+
+    /// Refresh Loop handler logic.
+    /// Should not return an error to prevent the actor from crashing.
+    async fn handle_refresh_loop(&mut self, ctx: &ActorContext<Self>) {
+        info!("retention-policy-refresh-indexes-operation");
+        self.counters.num_refresh_passes += 1;
+
+        let index_metadatas = match self.metastore.list_indexes_metadatas().await {
+            Ok(metadatas) => metadatas,
+            Err(error) => {
+                error!(error=?error, "Failed to list indexes from the metastore.");
+                return;
+            }
+        };
+        info!(index_ids=%index_metadatas.iter().map(|im| &im.index_id).join(", "), "Retention policy refresh.");
+
+        let deleted_indexes = compute_deleted_indexes(
+            self.index_metadatas.keys(),
+            index_metadatas.iter().map(|metadata| &metadata.index_id),
+        );
+        for index_id in &deleted_indexes {
+            self.index_metadatas.remove(index_id);
+        }
+        info!(index_ids=%deleted_indexes.iter().join(", "), "Deleted indexes from cache.");
+
+        for index_metadata in index_metadatas.into_iter() {
+            ctx.record_progress();
```

Comment:
Feels like this `record_progress` is useless. Everything in the loop happens in memory without long compute-intensive ops or IOs.